pkg/runtime/pubsub/cloudevents.go

```diff
@@ -14,28 +14,40 @@ limitations under the License.
 package pubsub
 
 import (
-	"github.com/google/uuid"
+	"github.com/mitchellh/mapstructure"
 
 	contribContenttype "github.com/dapr/components-contrib/contenttype"
 	contribPubsub "github.com/dapr/components-contrib/pubsub"
 )
 
 // CloudEvent is a request object to create a Dapr compliant cloudevent.
+// The cloud event properties can manually be overwritten by using metadata beginning with "cloudevent-" as prefix.
 type CloudEvent struct {
-	ID              string
-	Data            []byte
-	Topic           string
-	Pubsub          string
-	DataContentType string
-	TraceID         string
-	TraceState      string
+	ID              string `mapstructure:"cloudevent-id"`
+	Data            []byte `mapstructure:"-"` // cannot be overridden
+	Topic           string `mapstructure:"-"` // cannot be overridden
+	Pubsub          string `mapstructure:"-"` // cannot be overridden
+	DataContentType string `mapstructure:"-"` // cannot be overridden
+	TraceID         string `mapstructure:"cloudevent-traceid"`
+	TraceState      string `mapstructure:"cloudevent-tracestate"`
+	Source          string `mapstructure:"cloudevent-source"`
+	Type            string `mapstructure:"cloudevent-type"`
+	TraceParent     string `mapstructure:"cloudevent-traceparent"`
 }
 
 // NewCloudEvent encapsulates the creation of a Dapr cloudevent from an existing cloudevent or a raw payload.
-func NewCloudEvent(req *CloudEvent) (map[string]interface{}, error) {
+func NewCloudEvent(req *CloudEvent, metadata map[string]string) (map[string]interface{}, error) {
 	if contribContenttype.IsCloudEventContentType(req.DataContentType) {
 		return contribPubsub.FromCloudEvent(req.Data, req.Topic, req.Pubsub, req.TraceID, req.TraceState)
 	}
-	return contribPubsub.NewCloudEventsEnvelope(uuid.New().String(), req.ID, contribPubsub.DefaultCloudEventType,
+
+	// metadata beginning with "cloudevent-" are considered overrides to the cloudevent envelope
+	mapstructure.WeakDecode(metadata, &req) // allows ignoring of case
```

Comment:
Nit:

```suggestion
	_ = mapstructure.WeakDecode(metadata, &req) // allows ignoring of case
```

This way it's clear that you're ignoring the error on purpose

Reply:
Good call - let me see if I want to handle that error here.